src/Dialect/ONNX/Transforms/Recompose.cpp

```diff
@@ -340,6 +340,185 @@ struct RecomposeLayerNormFromMulPattern : public OpRewritePattern<ONNXMulOp> {
+
+    // Extract and validate weight tensor rank
+    auto weightType =
+        mlir::dyn_cast<RankedTensorType>(firstConv.getW().getType());
+    if (!weightType)
+      return failure();
+    int64_t rank = weightType.getRank();
+    if (1 >= rank)
+      return failure(); // Ensure axis is within valid range
+
+    // Ensure valid axis selection
+    int64_t concatAxis = 1; // Typically channel dimension
+    if (concatAxis >= rank)
+      return failure(); // Prevent out-of-range errors
+
+    // Get the number of split outputs
+    int64_t numSplits = splitOp.getNumResults();
+    if (numSplits != static_cast<int64_t>(convOps.size())) {
+      return failure();
+    }
+
+    // Ensure all ConvOps have the same kernel, stride, dilation, and padding
+    for (size_t i = 1; i < convOps.size(); ++i) {
+      if (convOps[i].getKernelShape() != firstConv.getKernelShape() ||
+          convOps[i].getStrides() != firstConv.getStrides() ||
+          convOps[i].getDilations() != firstConv.getDilations() ||
+          convOps[i].getPads() != firstConv.getPads()) {
+        return failure();
+      }
+    }
+
+    // Create correct IntegerAttrs
+    IntegerAttr axis0 = rewriter.getI64IntegerAttr(0);
+    IntegerType si64Type = rewriter.getIntegerType(64, /*isSigned=*/true);
+    IntegerAttr groupAttrVal =
+        IntegerAttr::get(si64Type, static_cast<int64_t>(numSplits));
+
+    // **Concatenating Conv Weights Correctly**
+    SmallVector<Value, 2> weightTensors;
+    int64_t total_C_out = 0;
+    for (auto conv : convOps) {
+      weightTensors.push_back(conv.getW());
+      auto wType = mlir::dyn_cast<RankedTensorType>(conv.getW().getType());
+      total_C_out += wType.getDimSize(0); // Summing output channels
+    }
+
+    // Compute correct concatenated shape
+    Type newWeightType = RankedTensorType::get(
+        {total_C_out, weightType.getDimSize(1), weightType.getDimSize(2),
+            weightType.getDimSize(3)},
+        weightType.getElementType());
+
+    // Create new concatenated weight tensor (concatenating along axis=0)
+    Location weightLoc =
+        firstConv.getW().getLoc(); // Get location from the first Conv weight
+    axis0 = IntegerAttr::get(si64Type, 0);
+
+    Value concatenatedWeight = rewriter.create<ONNXConcatOp>(
+        weightLoc, newWeightType, weightTensors, axis0);
+
+    // **Concatenating Bias Correctly**
+    SmallVector<Value, 2> biasTensors;
+    bool hasBias = llvm::all_of(convOps, [](ONNXConvOp conv) {
+      return conv.getB() && !mlir::isa<NoneType>(conv.getB().getType());
+    });
+
+    Value concatenatedBias = Value(); // Default empty Value
+    Location biasLoc =
+        hasBias ? firstConv.getB().getLoc()
+                : loc; // Get location from the first Conv bias if available
+    if (hasBias) {
+      for (auto conv : convOps)
+        biasTensors.push_back(conv.getB());
+
+      Type newBiasType =
+          RankedTensorType::get({total_C_out}, weightType.getElementType());
```

Comment:
Define `Type elementType = weightType.getElementType();` at the beginning of this function and reuse it to avoid boilerplate code.